test/core/tsi/crl_ssl_transport_security_test.cc

```diff
@@ -0,0 +1,280 @@
+// Copyright 2021 gRPC authors.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <grpc/grpc.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+
+#include "src/core/lib/iomgr/load_file.h"
+#include "src/core/lib/security/security_connector/security_connector.h"
+#include "src/core/tsi/ssl_transport_security.h"
+#include "src/core/tsi/transport_security.h"
+#include "src/core/tsi/transport_security_interface.h"
+#include "test/core/tsi/transport_security_test_lib.h"
+#include "test/core/util/test_config.h"
+
+extern "C" {
+#include <openssl/crypto.h>
+#include <openssl/pem.h>
+}
+
+namespace {
+
+const int kSslTsiTestRevokedKeyCertPairsNum = 1;
+const int kSslTsiTestValidKeyCertPairsNum = 1;
+const char* kSslTsiTestCrlSupportedCredentialsDir = "test/core/tsi/test_creds/";
+
+class CrlSslTransportSecurityTest
+    : public testing::TestWithParam<tsi_tls_version> {
+ protected:
+  // A tsi_test_fixture implementation.
+  class SslTsiTestFixture {
+   public:
+    static SslTsiTestFixture* Create(bool use_revoked_server_cert,
+                                     bool use_revoked_client_cert) {
+      auto* fixture = static_cast<SslTsiTestFixture*>(
+          gpr_malloc(sizeof(SslTsiTestFixture)));
+      new (fixture)
+          SslTsiTestFixture(use_revoked_server_cert, use_revoked_client_cert);
```

Comment:
I don't understand the code here. We first allocate the fixture with gpr_malloc, and then we immediately change its value with new? It also seems `fixture` is never released. Did I miss anything?
In general, fixture should be created either through `gpr_malloc` or `new`, not both. Since it's allocated dynamically, it should be freed by `gpr_free` or `delete`.

Reply:
This code is also very odd because if I add the following code to log the "client_result"/"server_result" inside of fixture->base_:

```
      auto* fixture = static_cast<SslTsiTestFixture*>(
          gpr_malloc(sizeof(SslTsiTestFixture)));
      std::string log = "fixture->base.client_result: ";
      log += fixture->base_.client_result == nullptr ? "nullptr" : "addr";
      log += ". fixture->base.server_result ";
      log += fixture->base_.server_result == nullptr ? "nullptr" : "addr";
      gpr_log(GPR_INFO, "%s", log.c_str());
      new (fixture)
          SslTsiTestFixture(use_revoked_server_cert, use_revoked_client_cert);
      return fixture;
```

It says that the client_result and server_results already have an address assigned to the pointers (only in ASAN mode). I believe this is what is causing the assertion error you found earlier @markdroth. I'm not really convinced this bug is a result of the CRL checking. It looks like a result of the unit test being converted to C++ to me.

Reply:
If I manually set the client/server results to nullptr the test passes:

```
      auto* fixture = static_cast<SslTsiTestFixture*>(
          gpr_malloc(sizeof(SslTsiTestFixture)));
      new (fixture)
          SslTsiTestFixture(use_revoked_server_cert, use_revoked_client_cert);
      fixture->base_.client_result = nullptr;
      fixture->base_.server_result = nullptr;
```

Reply:
I pulled the changes yesterday and did some investigations myself. I also had some discussions with @krestofur . Here is what we've found:

1. the error @markdroth saw was actually because **tsi_test_fixture** was originally created using **grpc_core::Zalloc**, and now we changed it to be the first member of a class, so we can't use **grpc_core::Zalloc** anymore, since the class should be trivial[1] 
2. I tried two ways, but neither of them gave me a clean result. I tried to use **gpr_malloc**, but it seems **gpr_malloc** won't initialize the class, but leave the member values undefined[2]. I tried to use **new** as well, but there were some memory leak as well(error logging: [3]).

To summarize, we want to use **grpc_core::Zalloc** but we couldn't, and we didn't find a suitable substitute, after changing the struct semantics to C++ semantics. To fundamentally solve this problem, I think the best way is to eliminate all underlying C semantics structs as well, but that will be a larger effort and is not very relevant to this PR.

I think I've had a better understanding about the the problem, just not sure what's our next step. Maybe we can have a meeting to discuss the possible solutions @krestofur @markdroth . Thanks!

[1] https://github.com/grpc/grpc/blob/master/src/core/lib/gprpp/memory.h#L51
[2] https://stackoverflow.com/questions/1538420/difference-between-malloc-and-calloc
[3] https://gist.github.com/ZhenLian/3bd2de8039504f43ba9fceadf2ecab58

Reply:
Update: I was able to completely get rid of that memory issue. See https://github.com/grpc/grpc/pull/27986  for a draft PR. 
But the other two tests still fail on the mismatch between some expected value and the real value. That seems to be the logic error. Passing to @krestofur to take a look. 
Feel free to let me know if there is anything I could help!